src/secp.rs

```diff
@@ -1,55 +1,99 @@
 pub use secp256k1zkp::{ContextFlag, Message, Secp256k1, Signature};
+pub use secp256k1zkp::aggsig;
 pub use secp256k1zkp::ecdh::SharedSecret;
 pub use secp256k1zkp::pedersen::{Commitment, RangeProof};
-pub use secp256k1zkp::key::{PublicKey, SecretKey};
+pub use secp256k1zkp::key::{PublicKey, SecretKey, ZERO_KEY};
 pub use secp256k1zkp::constants::{AGG_SIGNATURE_SIZE, COMPRESSED_PUBLIC_KEY_SIZE, MAX_PROOF_SIZE, PEDERSEN_COMMITMENT_SIZE, SECRET_KEY_SIZE};
 
-use crate::ser::{Readable, Reader, Writeable, Writer};
-use crate::error::{ErrorKind, Result};
+use crate::error::{Error, ErrorKind, Result};
 
-use rand::RngCore;
-use std::cmp;
+use blake2::blake2b::Blake2b;
+use byteorder::{BigEndian, ByteOrder};
+use grin_core::ser::{self, Readable, Reader, Writeable, Writer};
+use secp256k1zkp::rand::thread_rng;
 
 /// A generalized Schnorr signature with a pedersen commitment value & blinding factors as the keys
-pub const COM_SIGNATURE_SIZE : usize = 96;
+pub struct ComSignature {
+    pub_nonce: Commitment,
+    s: SecretKey,
+    t: SecretKey,
+}
 
-pub struct ComSignature(pub [u8; COM_SIGNATURE_SIZE]);
 impl ComSignature {
-    /// Builds a ComSignature from a byte vector. If the vector is too short, it will be
-    /// completed by zeroes. If it's too long, it will be truncated.
-    pub fn from_vec(v: Vec<u8>) -> ComSignature {
-        let mut h = [0; COM_SIGNATURE_SIZE];
-        for i in 0..cmp::min(v.len(), COM_SIGNATURE_SIZE) {
-            h[i] = v[i];
+    pub fn new(pub_nonce: &Commitment, s: &SecretKey, t: &SecretKey) -> ComSignature {

```

Comment:
nitpick: I think it would be nice to add a brief description of what this does, something along the lines of
```
P = v*H + r*G  # pubkey

# Sign message M with P
R = k1*H + k2*G             # Random nonce
e = H(M | P | R)            # Compute challenge
s, t = (e*v+k1), (e*r+k2)
sig = (s, t, R)             # This is our signature for M

# Verify signature of message M for P given (s, t, R)
e = H(M | P | R)            # Compute challenge
e*P + R = s*H + t*G         # Verify the equation
```
I've not checked this actually correctly reflects the code below, just an example of what I had in mind. It may make it easier for someone new reading the code to grasp the underlying idea.